File: ascii85.c
#include <stdio.h>

/* powers of 85 */

long divarray[5] = { (long) 1, (long) 85, (long) 7225, (long) 614125,
						   (long) 52200625 };
	 
#include <stdio.h>
#include <unistd.h>

int	vflag;
int	hflag;

extern void EncodeFile(void);

int
main(argc,argv)

int argc;
char **argv;

{
  int	c;
  
  while ((c = getopt(argc,argv,"vh")) != EOF) {
    switch (c) {
    case 'v':
      vflag++;
      break;
    case 'h':
      hflag++;
      break;
    }
  }
  if (hflag)
    fputs("%!\ncurrentfile /ASCII85Decode filter cvx exec\n",stdout);
  
  EncodeFile();
  return(0);
}

#define myputc(c,f,counter,maxcounter) (putc(c,f), (counter)++, ((counter) >= (maxcounter)) ? (putc('\n',f),counter=0) : 0)

void
EncodeFile()
{
  unsigned long base256 = 0;
  int phase = 0, theChar, nonzero, base85digit[5], index, nz;
  long inbytes = 0;
  long outbytes = 0;
  int	ochars = 0;
  int	maxochars = 76;
	 
  while((theChar = fgetc(stdin)) != EOF)
    {
      inbytes++;
      base256 += ((unsigned long)theChar) << (8 * (3 - phase++));
      
      if(phase == 4)
	{
	  for(index = 4; index >= 0; index--)
	    {
	      if((base85digit[index] = base256 / divarray[index]) != 0)
		{
		  nz = !0;
		}
	      base256 = base256 % divarray[index];
	    }
	  
	  if(nz)
	    {
	      for(index = 4; index >= 0; index--)
		{
		  myputc(base85digit[index] + '!', stdout,ochars,maxochars);
		  outbytes++;
		}
	    } else
	    {
	      myputc('z', stdout,ochars,maxochars);
	      outbytes++;
	    }
	  phase = 0;
	  base256 = 0;
	}
      
      if((inbytes & 0x0000FFFF) == 0x0000FFFF)
	{
	  if (vflag) {
	    fprintf(stderr, "\r%ld bytes in, %ld bytes out", inbytes, outbytes);
	    fflush(stderr);
	  }
	}
    }
  
  if(phase != 0)
    {
      for(index = 4; index >= (4-phase); index--)
	{
	  myputc(base256 / divarray[index] + '!', stdout,ochars,maxochars);
	  base256 = base256 % divarray[index];
	}
    }
  
  fprintf(stdout, "~>\n");
  
  fflush(stdout);
  return;
}

File: ascii85.py
#! /usr/bin/env python

import sys
import getopt
import os
import struct
from itertools import count,cycle

def usage(utyp, *msg):
    sys.stderr.write('Usage: %s\n' % os.path.split(sys.argv[0])[1])
    if msg:
        sys.stderr.write('Error: %s\n' % (repr(msg),))
    sys.exit(1)

class NWriter(object):
    def __init__(self,ofile,nchars):
        self.ofile = ofile
        self.nchars = nchars
        self.nn = 0
        return
    def write(self,s):
        while s or self.nn >= self.nchars:
            if self.nn >= self.nchars:
                self.ofile.write('\n')
                self.nn = 0
            s0,s = s[:self.nchars-self.nn],s[self.nchars-self.nn:]
            self.ofile.write(s0)
            self.nn += len(s0)
    def nl(self):
        if self.nn:
            self.ofile.write('\n')
            self.nn = 0

class Global:
    base85digits = ''.join(chr(i) for i in range(33,33+85))
    divarray = tuple(85**i for i in range(5))
    def __init__(self):
        self.vflag = 0
        self.hflag = 0
        return
    def encode(self,infile,outfile):
        unpack = struct.unpack
        ofl = NWriter(outfile,76)
        while True:
            d = infile.read(4)
            if not d:
                break
            lend = len(d)
            while len(d) < 4:
                d = d+b'\x00'
            base256 = unpack('>I',d)[0]
            digits = ''.join(self.base85digits[(base256 // self.divarray[i]) % 85] for i in range(4,-1,-1))
            ofl.write(digits[:lend+1])
        # ofl.nl()
        outfile.write('~>\n')
    def doit(self,args):
        if self.hflag:
            sys.stdout.buffer.write(b'%!\ncurrentfile /ASCII85Decode filter cvx exec\n')
        self.encode(sys.stdin.buffer,sys.stdout)
        return

def main(argv):
    gp = Global()
    try:
        opts, args = getopt.getopt(argv[1:],
                                   'hv',
                                   ['header',
                                    'verbose',
                                    ])
    except getopt.error as msg:
        usage(1, msg)

    for opt, arg in opts:
        if opt in ('-h', '--header'):
            gp.hflag += 1
        elif opt in ('-v', '--verbose'):
            gp.vflag += 1

    gp.doit(args)
        
if __name__ == '__main__':
    main(sys.argv)
